library/std/src/sys/unix/os.rs

```diff
@@ -131,6 +132,122 @@ pub fn error_string(errno: i32) -> String {
     }
 }
 
+// Signal strings
+//
+// This is not so easy.  Docs references:
+//
+// glibc
+//  https://www.sourceware.org/glibc/wiki/Release/2.32
+//  https://www.gnu.org/software/libc/manual/html_node/Signal-Messages.html
+//
+// FreeBSD
+//  https://www.freebsd.org/cgi/man.cgi?query=strsignal&apropos=0&sektion=0&manpath=FreeBSD+12.2-RELEASE+and+Ports&arch=default&format=html
+
+pub trait SignalLookupMethod {
+    fn lookup(&self, sig: i32) -> Option<&'static str>;
+}
+
+unsafe fn ptr_to_maybe_str(p: *const c_char) -> Option<&'static str> {
+    if p.is_null() {
+        None
+    } else {
+        let cstr = CStr::from_ptr::<'static>(p);
+        Some(cstr.to_str().unwrap())
+    }
+}
+
+// This is pretty universal - the nix crate makes the same assumption.  We need it to be no larger
+// than the platform's actual value - but, only on platforms where there is sys_siglist or
+// sys_signame.  This value is a pretty safe bet, and we have a test case for it.
+const NSIG: usize = 32;
+
+#[cfg(not(target_env = "musl"))]
+pub mod signal_lookup {
+    use super::*;
+    use crate::sys::weak::Weak;
+
+    type SigFooNp = unsafe extern "C" fn(c_int) -> *const c_char;
+
+    #[repr(transparent)]
+    struct SysSigFoos(*const *const c_char);
+    unsafe impl Sync for SysSigFoos {}
+
+    pub struct Lookups {
+        sigfoo_np: Weak<SigFooNp>,
+        sys_sigfoos: Weak<SysSigFoos>,
+    }
+
+    pub static descrs: Lookups = Lookups {
+        sigfoo_np: Weak::new("sigdescr_pn\0"),   // glibc >=2.32
+        sys_sigfoos: Weak::new("sys_siglist\0"), // FreeBSD/NetBSD/OpenBSD
+    };
+
+    pub static abbrevs: Lookups = Lookups {
+        sigfoo_np: Weak::new("sigabbrev_pn\0"),  // glibc >=2.32
+        sys_sigfoos: Weak::new("sys_signame\0"), // glibc < 2.23, BSDs, and other trad unix
+    };
+
+    impl SignalLookupMethod for Lookups {
+        fn lookup(&self, sig: i32) -> Option<&'static str> {
+            unsafe {
+                let got = if let Some(sigfoo_np) = self.sigfoo_np.get() {
+                    sigfoo_np(sig)
+                } else if let Some(use_sys_foolist) =
+                    // if let chaining would be nice here!
+                    if (sig as usize) < NSIG { self.sys_sigfoos.get() } else { None }
+                {
+                    use_sys_foolist.0.add(sig as usize).read()
+                } else {
+                    ptr::null()
+                };
+                ptr_to_maybe_str(got)
+            }
+        }
+    }
+}
+
+#[cfg(target_env = "musl")]
+pub mod signal_lookup {
+    pub struct ViaStrsignal;
+    pub static descrs: ViaStrSignal = ViaStrSignal;
+    impl Lookup for ViaStrSignal {
+        // musl's strsignal is threadsafe: it just looks up in a static array (that we don't have
+        // access to in any other way).  I have spoken to Rich Felker (musl upstream author) on IRC
+        // and confirmed that musl's strsignal is very unlikely to ever become non-thread-safe.
```

Comment:
I agree that having this commitment documented would be better.  But as far as I can tell musl does not have the kind of comprehensive documentation where this kind of thing would go.  Otherwise I would have tried that already.  Also AFAICT there isn't an issue tracker.  I guess I could post to the mailing list.  Would that satsify this concern?

I doubt very much that musl upstream would want to provide `sigdescr_np`.  They focus heavily on standards conformance and `sigdescr_np` is an API glibc simply made up (understandably so, but, still).

Reply:
> It looks like musl has fixed strings for all the real-time signals. Is musl's handling of real-time and unknown signals included in the above commitment?

My conversation was about the return value from `strsignal` and did not distinguish the behaviour depending on the argument values.  Such a distinction wouldn't make sense as an API commitment since the caller couldn't rely on the threadsafety.

Reply:
A public mailing list post with an ack from the maintainer would be helpful, yes.

And I didn't want to make any assumptions about the scope of that informal commitment; I could have imagined something like "if you pass a known signal number we'll guarantee to return a static, but we might change the behavior for unknown signal numbers in the future". That's still a commitment we could rely on, as long as we were careful to only pass known signal numbers. Hence trying to confirm the scope of the commitment.